tianshou/data/batch.py

```diff
@@ -114,17 +118,20 @@ def _assert_type_keys(keys):
 
 def _parse_value(v: Any):
     if isinstance(v, (list, tuple, np.ndarray)):
-        if not isinstance(v, np.ndarray) and \
-                all(isinstance(e, torch.Tensor) for e in v):
-            v = torch.stack(v)
-            return v
-        v_ = _to_array_with_correct_type(v)
-        if v_.dtype == np.object and _is_batch_set(v):
-            v = Batch(v)  # list of dict / Batch
-        else:
-            # normal data list (main case)
-            # or actually a data list with objects
-            v = v_
+        try:
+            if not isinstance(v, np.ndarray) and \
+                    all(isinstance(e, torch.Tensor) for e in v):
+                return torch.stack(v)
+            v_ = _to_array_with_correct_type(v)
+            if v_.dtype == np.object and _is_batch_set(v):
+                v = Batch(v)  # list of dict / Batch
+            else:
+                # normal data list (main case)
+                # or actually a data list with objects
+                v = v_
+        except (ValueError, RuntimeError):
+            raise TypeError("Batch does not support non-stackable list/tuple "
+                            "of tensors as value yet.")
```

Comment:
Somehow I'm not comfortable with such a large try-catch block. Maybe we can just leave the exceptions as they are. Such a large try-catch block may catch some unintended exceptions and mislead users.

Reply:
> Finally, it does nothing more than raising an exception, so the trace is preserved.

My point is that, not only the trace but also the exception message should remain, so that users know what causes the exception. For example, torch.cat gives message like "RuntimeError: stack expects each tensor to be equal size, but got [3, 4] at entry 0 and [3, 5] at entry 1", this message is valuable for debugging, but will be replaced by "Batch does not support non-stackable list/tuple of tensors as value yet." in this pr.

Reply:
> "RuntimeError: stack expects each tensor to be equal size, but got [3, 4] at entry 0 and [3, 5] at entry 1", this message is valuable for debugging, but will be replaced by "Batch does not support non-stackable list/tuple of tensors as value yet." in this pr.

So it is basically the same but less specific.

Reply:
> I can forward the message to ease debugging.

Done !

Reply:
> I can forward the message to ease debugging.

With this, I'm ok now. You can merge as you like. Anyway, I still don't get the point of this big try-catch block.

Reply:
Ok I will split it if it is clearer for you. No problem.

Reply:
If we are to collaborate, we have to make sure everybody understands and is ok with each other's PR.

Reply:
> Ok I will split it if it is clearer for you. No problem.

It looks much better if you just catch the ``v_ = _to_array_with_correct_type(v)`` statement or do it inside ``_to_array_with_correct_type``. It does not make sense to catch other statements, I think.

```python
        try:
            if not isinstance(v, np.ndarray) and \
                    all(isinstance(e, torch.Tensor) for e in v):
                return torch.stack(v) # no need to catch here, users clearly know what caused the error when an exception happens here
            v_ = _to_array_with_correct_type(v)
            if v_.dtype == np.object and _is_batch_set(v):
                # no need to catch here, users clearly know what caused the error when an exception happens here
                v = Batch(v)  # list of dict / Batch
            else:
                # normal data list (main case)
                # or actually a data list with objects
                v = v_ # no need to catch here, users clearly know what caused the error when an exception happens here
        except (ValueError, RuntimeError):
            raise TypeError("Batch does not support non-stackable list/tuple "
                            "of tensors as value yet.")
```

Reply:
It should be fine now. Let me know.